src/main/scala/TableGroup.scala

```diff
@@ -1,6 +1,412 @@
-package typeDefinitions
-case class TableGroup(
-                     id: String, // Int?
-                     notes: String, // revisit
-                     tables: List[Table]
-                     )
\ No newline at end of file
+package CSVValidation
+
+import CSVValidation.traits.JavaIteratorExtensions.IteratorHasAsScalaArray
+import CSVValidation.traits.ObjectNodeExtentions.IteratorHasGetKeysAndValues
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{
+  ArrayNode,
+  JsonNodeFactory,
+  ObjectNode,
+  TextNode
+}
+import errors.MetadataError
+
+import scala.collection.mutable.Map
+import java.net.{URI, URL}
+import scala.collection.mutable
+
+object TableGroup {
+  val csvwContextUri = "http://www.w3.org/ns/csvw"
+  val validProperties: Array[String] = Array[String]("tables", "notes", "@type")
+
+  def fromJson(tableGroupNode: ObjectNode, baseUri: String): TableGroup = {
+    var baseUrl = baseUri.trim
+    val containsWhitespaces = ".*\\s.*".r
+    val matcher = containsWhitespaces.pattern.matcher(baseUrl)
+    if (matcher.matches()) {
+      println(
+        "Warning: The path/url has whitespaces in it, please ensure its correctness. Proceeding with received " +
+          "path/url .."
+      )
+    }
+    val processContextResult =
+      processContextGetBaseUrlLang(tableGroupNode, baseUrl, "und")
+    baseUrl = processContextResult._1
+    val lang = processContextResult._2
+
+    restructureIfNodeIsSingleTable(tableGroupNode)
+
+    var (annotations, commonProperties, inheritedProperties, warnings) =
+      classifyPropertiesBasedOnPropertyTypeAndSetWarnings(
+        tableGroupNode,
+        baseUrl,
+        lang
+      )
+
+    val id = getId(commonProperties)
+    ensureTypeofTableGroup(tableGroupNode)
+
+    val (tables, w) = createTableObjectsAndSetWarnings(
+      tableGroupNode,
+      baseUrl,
+      lang,
+      commonProperties,
+      inheritedProperties
+    )
+    warnings = warnings.concat(w)
+
+    findForeignKeysLinkToReferencedTables(baseUrl, tables)
+
+    return TableGroup(
+      baseUrl,
+      id,
+      tables,
+      getNotes(commonProperties),
+      annotations,
+      warnings
+    )
+  }
+
+  private def restructureIfNodeIsSingleTable(
+      tableGroupNode: ObjectNode
+  ): Unit = {
+    val tableGroupNodeCopy = tableGroupNode.deepCopy()
```

Comment:
I think I need it here, otherwise I get an exception from Jackson.
`An exception or error caused a run to abort. 
java.lang.StackOverflowError
	at com.fasterxml.jackson.core.JsonStreamContext.<init>(JsonStreamContext.java:60)
	at com.fasterxml.jackson.core.json.JsonWriteContext.<init>(JsonWriteContext.java:79)
	at com.fasterxml.jackson.core.json.JsonWriteContext.createChildObjectContext(JsonWriteContext.java:194)
	at com.fasterxml.jackson.core.json.WriterBasedJsonGenerator.writeStartObject(WriterBasedJsonGenerator.java:327)
	at com.fasterxml.jackson.databind.node.ObjectNode.serialize(ObjectNode.java:312)`

Reply:
Shall we pair up in a bit and take a look at this? I'm a bit concerned if we're getting that.